p2p/discover/v4_udp.go

```diff
@@ -750,7 +831,41 @@ func decodeV4(buf []byte) (packetV4, encPubkey, []byte, error) {
 	return req, fromKey, hash, err
 }
 
-// Packet Handlers
+// checkBond checks if the given node has a recent enough endpoint proof.
+func (t *UDPv4) checkBond(id enode.ID, ip net.IP) bool {
+	return time.Since(t.db.LastPongReceived(id, ip)) < bondExpiration
+}
+
+// ensureBond solicits a ping from a node if we haven't seen a ping from it for a while.
+// This ensures there is a valid endpoint proof on the remote end.
+func (t *UDPv4) ensureBond(toid enode.ID, toaddr *net.UDPAddr) {
+	tooOld := time.Since(t.db.LastPingReceived(toid, toaddr.IP)) > bondExpiration
+	if tooOld || t.db.FindFails(toid, toaddr.IP) > 5 {
```

Comment:
Please separate out this `5` into it's own constant. Also, previously we didn't seem care about the number of failures. Any reason for asserting a 5 here? What's the rational? What's the expected behavioral difference?

Reply:
I just reused the existing maxFindnodeFailures constant now which was already five. There are two important differences in the new handling of findnode failures:

- we didn't reset the counter value but reduced it by one for every successful reply. If a remote node was unreachable for a while, the counter would go up to a very high value and it would take forever to come down below five again. 
- the counter was not used for rebonding, i.e. if a node didn't respond to findnode we never even tried pinging it again. Doing ping/pong again in this case will make communication work if the node is actually live but just forgot about our endpoint proof.